sequencer/src/state.rs

```diff
@@ -1,47 +1,155 @@
+        let fee_merkle_tree_root = fee_merkle_tree.commitment();
+        anyhow::ensure!(
+            parent.fee_merkle_tree_root == fee_merkle_tree_root,
+            anyhow::anyhow!(
+                "Invalid Fee Merkle Tree Error: {}, {}",
+                parent.fee_merkle_tree_root,
+                fee_merkle_tree_root
+            )
+        );
+
+        // validate parent block merkle tree root against state
+        // `clone` to avoid possible side-effects
+        let mut block_merkle_tree = self.block_merkle_tree.clone();
+        let block_merkle_tree_root = block_merkle_tree.commitment();
+        anyhow::ensure!(
+            parent.block_merkle_tree_root == block_merkle_tree_root,
+            anyhow::anyhow!(
+                "Invalid Block Merkle Tree Error: {}, {}",
+                parent.block_merkle_tree_root,
+                block_merkle_tree_root
+            )
+        );
+
+        // validate proposal is descendent of parent by appending to parent
+        block_merkle_tree.push(parent.commit()).unwrap();
+        let block_merkle_tree_root = block_merkle_tree.commitment();
+        anyhow::ensure!(
+            proposal.block_merkle_tree_root == block_merkle_tree_root,
+            anyhow::anyhow!(
+                "Invalid Block Merkle Tree Error: {}, {}",
+                block_merkle_tree_root,
+                proposal.block_merkle_tree_root
+            )
+        );
+
+        Ok(ValidatedState {
+            block_merkle_tree,
+            fee_merkle_tree,
+        })
+    }
+}
+
+impl HotShotState for ValidatedState {
+    type Error = BlockError;
+    type Instance = NodeState;
     type BlockHeader = Header;
     type BlockPayload = Payload;
 
     type Time = ViewNumber;
 
     fn on_commit(&self) {}
-
+    /// Validate parent against known values (from state) and validate
+    /// proposal descends from parent. Returns updated `ValidatedState`.
     fn validate_and_apply_header(
         &self,
-        _proposed_header: &Self::BlockHeader,
-        _parent_header: &Self::BlockHeader,
-        _view_number: &Self::Time,
+        _instance: &Self::Instance,
+        parent_header: &Self::BlockHeader,
+        proposed_header: &Self::BlockHeader,
     ) -> Result<Self, Self::Error> {
-        todo!("Tracking issue: [https://github.com/EspressoSystems/espresso-sequencer/issues/968]")
+        match self.validate_proposal(parent_header, proposed_header) {
+            // Note that currently only block state is updated.
+            Ok(validated_state) => Ok(validated_state),
+            Err(e) => {
+                tracing::warn!("Invalid Proposal: {}", e);
+                Err(BlockError::InvalidBlockHeader)
+            }
+        }
+    }
+    /// Construct the state with the given block header.
+    ///
+    /// This can also be used to rebuild the state for catchup.
+    fn from_header(_block_header: &Self::BlockHeader) -> Self {
+        ValidatedState::default()
```

Comment:
It will be hard to tell if the tests are working with this dummy implementation in. I'm a bit afraid we might updating tests in ways that they shouldn't be updated, and then have to fix them again once we implement this for real. So I think it might be a good idea to go ahead and implement this now. You can use [`from_commitment`](https://jellyfish.docs.espressosys.com/jf_primitives/merkle_tree/trait.ForgetableMerkleTreeScheme.html#tymethod.from_commitment) to generate completely sparse views of the merkle trees from teh commitments contained in `block_header`

Reply:
fixed in c8440bfff33f85094ae1635a4c0761a4be741527